File: src/stflrichtext.cpp
#include "stflrichtext.h"
#include "libnewsboat-ffi/src/stflrichtext.rs.h"

namespace newsboat {

StflRichText::StflRichText(rust::Box<stflrichtext::bridged::StflRichText>&& rs_object)
	: rs_object(std::move(rs_object))
{
}

StflRichText::StflRichText(const StflRichText& other)
	: rs_object(stflrichtext::bridged::copy(*other.rs_object))
{
}

StflRichText& StflRichText::operator=(const StflRichText& other)
{
	this->rs_object = stflrichtext::bridged::copy(*other.rs_object);
	return *this;
}

StflRichText StflRichText::from_plaintext(const std::string& text)
{
	auto rs_object = stflrichtext::bridged::from_plaintext(text);

	return StflRichText(std::move(rs_object));
}

StflRichText StflRichText::from_plaintext_with_style(const std::string& text,
	const std::string& style_tag)
{
	auto rs_object = stflrichtext::bridged::from_plaintext_with_style(text, style_tag);

	return StflRichText(std::move(rs_object));
}

StflRichText StflRichText::from_quoted(const std::string& text)
{
	auto rs_object = stflrichtext::bridged::from_quoted(text);

	return StflRichText(std::move(rs_object));
}

void StflRichText::append(const StflRichText& other)
{
	stflrichtext::bridged::append(*rs_object, *other.rs_object);
}

void StflRichText::highlight_searchphrase(const std::string& search, bool case_insensitive)
{
	stflrichtext::bridged::highlight_searchphrase(*rs_object, search, case_insensitive);
}

void StflRichText::apply_style_tag(const std::string& tag, size_t start, size_t end)
{
	stflrichtext::bridged::apply_style_tag(*rs_object, tag, start, end);
}

std::string StflRichText::plaintext() const
{
	return std::string(stflrichtext::bridged::plaintext(*rs_object));
}

std::string StflRichText::stfl_quoted() const
{
	return std::string(stflrichtext::bridged::quoted(*rs_object));
}

}

File: rust/libnewsboat-ffi/src/stflrichtext.rs
use cxx::CxxString;
use libnewsboat::stflrichtext;

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
struct StflRichText(stflrichtext::StflRichText);

#[cxx::bridge(namespace = "newsboat::stflrichtext::bridged")]
mod ffi {
    extern "Rust" {
        type StflRichText;

        fn from_plaintext(text: &CxxString) -> Box<StflRichText>;
        fn from_plaintext_with_style(text: &CxxString, style_tag: &str) -> Box<StflRichText>;
        fn from_quoted(text: &CxxString) -> Box<StflRichText>;
        fn copy(richtext: &StflRichText) -> Box<StflRichText>;
        fn append(richtext: &mut StflRichText, other: &StflRichText);

        fn highlight_searchphrase(
            richtext: &mut StflRichText,
            search: &str,
            case_insensitive: bool,
        );
        fn apply_style_tag(richtext: &mut StflRichText, tag: &str, start: usize, end: usize);
        fn plaintext(richtext: &StflRichText) -> &str;
        fn quoted(richtext: &StflRichText) -> String;
    }
}

fn from_plaintext(text: &CxxString) -> Box<StflRichText> {
    let text = text.to_string_lossy();
    let richtext = stflrichtext::StflRichText::from_plaintext(&text);
    Box::new(StflRichText(richtext))
}

fn from_plaintext_with_style(text: &CxxString, style_tag: &str) -> Box<StflRichText> {
    let text = text.to_string_lossy();
    let richtext = stflrichtext::StflRichText::from_plaintext_with_style(&text, style_tag);
    Box::new(StflRichText(richtext))
}

fn from_quoted(text: &CxxString) -> Box<StflRichText> {
    let text = text.to_string_lossy();
    let richtext = stflrichtext::StflRichText::from_quoted(&text);
    Box::new(StflRichText(richtext))
}

fn copy(richtext: &StflRichText) -> Box<StflRichText> {
    Box::new(StflRichText(richtext.0.clone()))
}

fn append(richtext: &mut StflRichText, other: &StflRichText) {
    richtext.0.append(&other.0);
}

fn highlight_searchphrase(richtext: &mut StflRichText, search: &str, case_insensitive: bool) {
    richtext.0.highlight_searchphrase(search, case_insensitive);
}

fn apply_style_tag(richtext: &mut StflRichText, tag: &str, start: usize, end: usize) {
    richtext.0.apply_style_tag(tag, start, end);
}

fn plaintext(richtext: &StflRichText) -> &str {
    richtext.0.plaintext()
}

fn quoted(richtext: &StflRichText) -> String {
    richtext.0.quoted()
}
